src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+package pl.edu.agh.firecell.engine.algorithm;
+
+import org.joml.Vector3i;
+import pl.edu.agh.firecell.model.Cell;
+import pl.edu.agh.firecell.model.State;
+import pl.edu.agh.firecell.model.util.NeighbourUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static pl.edu.agh.firecell.engine.algorithm.FirePropagator.isCellBurning;
+
+public class SmokePropagator {
+
+    private final double deltaTime;
+
+    public SmokePropagator(double deltaTime) {
+        this.deltaTime = deltaTime;
+    }
+
+    public double computeNewSmokeIndicator(State oldState, Vector3i cellIndex, Cell oldCell) {
+        int smokeFromFire = 0;
+        if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.down(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).isSolid()) {
+            smokeFromFire = oldState.getCell(NeighbourUtils.down(cellIndex)).material().smokeCoe();
+        }
+        Optional<Cell> northCell = Optional.empty();
+        Optional<Cell> southCell = Optional.empty();
+        Optional<Cell> westCell = Optional.empty();
+        Optional<Cell> eastCell = Optional.empty();
+        if (oldState.hasCell(NeighbourUtils.west(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.west(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.west(cellIndex)).isSolid()) {
+            westCell = Optional.of(oldState.getCell(NeighbourUtils.west(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.east(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.east(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.east(cellIndex)).isSolid()) {
+            eastCell = Optional.of(oldState.getCell(NeighbourUtils.east(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.north(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.north(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.north(cellIndex)).isSolid()) {
+            northCell = Optional.of(oldState.getCell(NeighbourUtils.north(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.south(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.south(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.south(cellIndex)).isSolid()) {
+            southCell = Optional.of(oldState.getCell(NeighbourUtils.south(cellIndex)));
+        }
+        if (southCell.isPresent()) smokeFromFire += southCell.get().material().smokeCoe() / 4;
+        if (northCell.isPresent()) smokeFromFire += northCell.get().material().smokeCoe() / 4;
+        if (westCell.isPresent()) smokeFromFire += westCell.get().material().smokeCoe() / 4;
+        if (eastCell.isPresent()) smokeFromFire += eastCell.get().material().smokeCoe() / 4;
```

Comment:
This is totally repetitive code, try to put it into some stream or list. If you can't, that's fine and leave it, let's not over-engeneer that.